### PR TITLE
Bandwidth zero

### PR DESCRIPTION
**Pull request by ptdatta**
What: Issue #5701 

Why: 
The nodes can have five types of status:
1. Online 
2. Offline
3. Not reachable
4. Unauthorized
5. Internal Error

However, the frontend is only configured to check the first 3 types.

The server is trying to check all types of node usage however the nodes that are not online, it tries to check their usage too which causes the error. 

Fix:
If any node is not online then just skip that.

Please describe the tests:
 - Test 1:
 - Test 2:
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [ ] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/main/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/main/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

### CODE REVIEW

**Comment by AlexeyALeonov** [comment]
@ptdatta you need to sign CLA as I can see, and rebase this PR

**Comment by ptdatta** [comment]
> @ptdatta you need to sign CLA as I can see, and rebase this PR

I signed it two times actually before making the PR.

File: multinode/nodes/service.go [hunk 1]
```diff
@@ -315,6 +318,10 @@ func (service *Service) TrustedSatellites(ctx context.Context) (_ storj.NodeURLs
 
 	var trustedSatellites storj.NodeURLs
 	for _, node := range listNodes {
+		nodeinfo := service.FetchNodeInfo(ctx, node)
```
Comment (ifraixedes) [inline multinode/nodes/service.go:321]: We would also use here the function that checks the status of the node

File: multinode/bandwidth/service.go [hunk 2]
```diff
@@ -105,12 +105,15 @@ func (service *Service) MonthlySatellite(ctx context.Context, satelliteID storj.
 	cache := make(UsageRollupDailyCache)
 
 	for _, node := range listNodes {
+		nodeinfo := service.nodes.FetchNodeInfo(ctx, node)
```
Comment (ifraixedes) [inline multinode/bandwidth/service.go:108]: It's a bit overkill to fetch all the node info only to check if it's online.
We could track the check from `FetchNodeInfo` to a separate function and call it from there and here.

File: multinode/storage/service.go [hunk 3]
```diff
@@ -126,6 +126,10 @@ func (service *Service) TotalUsageSatellite(ctx context.Context, satelliteID sto
 	cache := make(UsageStampDailyCache)
 
 	for _, node := range nodesList {
+		nodeinfo := service.nodes.FetchNodeInfo(ctx, node)
```
Comment (ifraixedes) [inline multinode/storage/service.go:129]: The same that I commented before, we would also use here the function that checks the status of the node.

File: multinode/storage/service.go [hunk 4]
```diff
@@ -159,6 +163,10 @@ func (service *Service) TotalDiskSpace(ctx context.Context) (totalDiskSpace Disk
 	}
 
 	for _, node := range listNodes {
+		nodeinfo := service.nodes.FetchNodeInfo(ctx, node)
```
Comment (ifraixedes) [inline multinode/storage/service.go:166]: The same that I commented before, we would also use here the function that checks the status of the node.

File: web/multinode/src/app/store/bandwidth.ts [hunk 5]
```diff
@@ -55,9 +56,8 @@ export class BandwidthModule implements Module<BandwidthState, RootState> {
      * @param ctx - context of the Vuex action.
      */
     public async fetch(ctx: ActionContext<BandwidthState, RootState>): Promise<void> {
-        const selectedSatelliteId = ctx.rootState.nodes.selectedSatellite ? ctx.rootState.nodes.selectedSatellite.id : null;
+        const selectedSatelliteId = ctx.rootState.nodes.selectedSatellite ? ctx.rootState.nodes.selectedSatellite.id : ctx.rootState.nodes.trustedSatellites[0].id;
```
Comment (ifraixedes) [inline web/multinode/src/app/store/bandwidth.ts:59]: Can we guarantee that always `trustedSatellites` will have at least one element?

File: web/multinode/src/app/store/storage.ts [hunk 6]
```diff
@@ -67,7 +67,7 @@ export class StorageModule implements Module<StorageState, RootState> {
      * @param ctx - context of the Vuex action.
      */
     public async usage(ctx: ActionContext<StorageState, RootState>): Promise<void> {
-        const selectedSatelliteId = ctx.rootState.nodes.selectedSatellite ? ctx.rootState.nodes.selectedSatellite.id : null;
+        const selectedSatelliteId = ctx.rootState.nodes.selectedSatellite ? ctx.rootState.nodes.selectedSatellite.id : ctx.rootState.nodes.trustedSatellites[0].id;
```
Comment (ifraixedes) [inline web/multinode/src/app/store/storage.ts:70]: Can we guarantee that always trustedSatellites will have at least one element?

File: web/multinode/src/nodes/index.ts [hunk 7]
```diff
@@ -49,7 +51,7 @@ export class Node {
     }
 
     public get statusText(): string {
-        return NodeStatus[this.status];
+        return this.status;
```
Comment (ifraixedes) [inline web/multinode/src/nodes/index.ts:54]: Why this change?

Comment (ptdatta) [reply web/multinode/src/nodes/index.ts:54]: @ifraixedes when we are fetching nodes we are already setting the status to `NodeStatus[node.status]`.
```
 return nodeListJson.map(node => new Node(
            node.id,
            node.name,
            node.version,
            new Date(node.lastContact),
            node.diskSpaceUsed,
            node.diskSpaceLeft,
            node.bandwidthUsed,
            0,
            0,
            0,
            node.totalEarned,
            NodeStatus[node.status],
        ));```
say if the node status is `online` from the api then in the node object this.status become `NodeStatus["online"]` which is `Online`.  So for `NodeStatus[this.status]` it will give `NodeStatus['Online']` which results `undefined`.

File: multinode/nodes/service.go (hunk 1 above)
Comment (ptdatta) [reply multinode/nodes/service.go:321]: @ifraixedes the nodestatus function only checks where the node is `online` or `offline` based on it's last contract. But the nodes can have other status also I think I should create a separate function to check the status of the node and use it every where.

File: web/multinode/src/app/store/bandwidth.ts (hunk 5 above)
Comment (ptdatta) [reply web/multinode/src/app/store/bandwidth.ts:59]: If there is a node there should be a satellite. Let me check so that I can make it better.

File: multinode/storage/service.go [hunk 8]
```diff
@@ -159,6 +163,10 @@ func (service *Service) TotalDiskSpace(ctx context.Context) (totalDiskSpace Disk
...
+		if nodeinfo.Status != nodes.StatusOnline {
+			continue
+		}
 		diskSpace, err := service.dialDiskSpace(ctx, node)
 		if err != nil {
```
Comment (ptdatta) [inline multinode/storage/service.go:171]: @ifraixedes we can also do one things if there is a err then we can just skip that. What so you think?

File: multinode/nodes/service.go (hunk 1 above)
Comment (ifraixedes) [reply multinode/nodes/service.go:321]: > I think I should create a separate function to check the status of the node and use it every where

That's what I wanted to say in my comment https://github.com/storj/storj/pull/7146/files#r1768432738

And use it here and in other places where I dropped the same comment.

File: multinode/storage/service.go (hunk 8 above)
Comment (ifraixedes) [reply multinode/storage/service.go:171]: I don't understand well your question.

What do you suggest to do if there is an error?

File: web/multinode/src/app/store/bandwidth.ts (hunk 5 above)
Comment (ifraixedes) [reply web/multinode/src/app/store/bandwidth.ts:59]: I see, my fault!

File: web/multinode/src/nodes/index.ts (hunk 7 above)
Comment (ifraixedes) [reply web/multinode/src/nodes/index.ts:54]: I'm not much into the frontend, but in your above code snipped, it uses `node.status`, no `node.statusText()`

Comment (ptdatta) [reply web/multinode/src/nodes/index.ts:54]: It actually uses 
```export enum NodeStatus {
    'online' = 'Online',
    'offline' = 'Offline',
    'not reachable' = 'Not reachable',
    'unauthorized' = 'Unauthorized',
    'storagenode internal error' = 'Internal Error'
}``` 
This enum. where status is the key and statusText is the value. That's why I think it was named like this. 
Without the change the statusText will return NodeStatus['Online'] which is undefined.

File: multinode/storage/service.go (hunk 8 above)
Comment (ptdatta) [reply multinode/storage/service.go:171]:
```
monthly, err := service.getMonthly(ctx, node)
		if err != nil {
			continue
		}
``` 
in stead of 
```
monthly, err := service.getMonthly(ctx, node)
		if err != nil {
			if nodes.ErrNodeNotReachable.Has(err) {
				continue
			}

			return Monthly{}, Error.Wrap(err)
		}
``` We can just skip the nodes which have errors.

File: multinode/nodes/service.go (hunk 1 above)
Comment (ptdatta) [reply multinode/nodes/service.go:321]: Oops my bad. I thought you are talking about the nodeStatus function. Sorry. 
I am creating a new one.

**Comment by ptdatta** [comment]
@ifraixedes I fetched `version` and `lastcontact` of the node with `status` as if I don't then I have to call the `nodeClient.Version` and `nodeClient.LastContact` twice

File: web/multinode/src/nodes/index.ts (hunk 7 above)
Comment (ptdatta) [reply web/multinode/src/nodes/index.ts:54]: It returns the status on the node according to NodeStatus enum values. The keys are all in lowercase but the values are capitalized.

Comment (profclems) [reply web/multinode/src/nodes/index.ts:54]: @ptdatta is right. In the constructor, the `this.status` is already set to `NodeStatus[node.status]`. Calling that again in `statusText` will most likely return an empty string.

Comment (ptdatta) [reply web/multinode/src/nodes/index.ts:54]: Yes it's returning `undefined`.

File: multinode/nodes/service.go [hunk 9]
```diff
@@ -368,15 +332,50 @@ func (service *Service) trustedSatellites(ctx context.Context, node Node) (_ sto
...
+		ApiKey: node.APISecret[:],
+	}
+
+	nodeVersion, err := nodeClient.Version(ctx, &multinodepb.VersionRequest{Header: header})
+	if err != nil {
+		if rpcstatus.Code(err) == rpcstatus.Unauthenticated {
+			return StatusUnauthorized, nil,nil
+		}
+
+		return StatusStorageNodeInternalError, nil,nil
```
Comment (profclems) [inline multinode/nodes/service.go:362]: Like I said above, we can still log the error and return the Internal server status

File: multinode/storage/service.go (hunk 8 above)
Comment (profclems) [reply multinode/storage/service.go:175]: I agree with skipping the nodes with errors as long as the status shows that they're not online. Also we shouldn't entirely ignore the errors but at least log them

**Review by profclems — CHANGES_REQUESTED** [review]
This is a really great patch. Thanks @ptdatta

File: web/multinode/src/nodes/index.ts (hunk 7 above)
Comment (ifraixedes) [reply web/multinode/src/nodes/index.ts:54]: My question was "What is calling this method". If it turns out that nothing calls it, could we remove it?

File: web/multinode/src/app/store/nodes.ts [hunk 10]
```diff
@@ -99,6 +100,18 @@ export class NodesModule implements Module<NodesState, RootState> {
...
     }
 
+    /**
+     * fetch action loads only online nodes information.
+     * @param ctx - context of the Vuex action.
+     */
+    public async fetchOnline(ctx: ActionContext<NodesState, RootState>): Promise<void> {
+        const nodes = ctx.state.selectedSatellite ? await this.nodes.listBySatellite(ctx.state.selectedSatellite.id) : await this.nodes.list();
+        
+        const onlineNodes = nodes.filter((node: Node) => node.statusText === NodeStatus['online']);
```
Comment (ptdatta) [inline web/multinode/src/app/store/nodes.ts:110]: @ifraixedes Here the `statusText` func is called.  I created this function because in the node's dropdown at the bandwidth page we only need the nodes which are online.

File: multinode/nodes/service.go [hunk 11]
```diff
@@ -368,15 +332,50 @@ func (service *Service) trustedSatellites(ctx context.Context, node Node) (_ sto
...
 }
 
-// nodeStatus chooses node status offline or online depends on LastContact.
-func nodeStatus(lastContact time.Time) Status {
-	now := time.Now().UTC()
 
-	if now.Sub(lastContact) < time.Hour*3 {
-		return StatusOnline
+// Fetch meta information about a node status, version, and last contact.
+func (service *Service) FetchNodeMeta(ctx context.Context, node Node) (_ Status,_*multinodepb.VersionResponse,_ *multinodepb.LastContactResponse) {
```
[the remark this answers is not in the document]
Comment (ptdatta) [reply multinode/nodes/service.go:337]: @profclems Logs added

File: web/multinode/src/app/store/nodes.ts (hunk 10 above)
Comment (ifraixedes) [reply web/multinode/src/app/store/nodes.ts:110]: Is it not possible to use `node.status` directly?

Comment (ptdatta) [reply web/multinode/src/app/store/nodes.ts]: Yes, it's possible. I removed the `statusText` method.

**Review by ifraixedes — DISMISSED** [review]

File: web/multinode/src/app/store/nodes.ts [hunk 12]
```diff
@@ -106,9 +106,13 @@ export class NodesModule implements Module<NodesState, RootState> {
      */
     public async fetchOnline(ctx: ActionContext<NodesState, RootState>): Promise<void> {
         const nodes = ctx.state.selectedSatellite ? await this.nodes.listBySatellite(ctx.state.selectedSatellite.id) : await this.nodes.list();
-        
-        const onlineNodes = nodes.filter((node: Node) => node.status === NodeStatus['online']);
-        
+        let onlineNodes;
+        if (nodes || Array.isArray(nodes)) {
```
Comment (ifraixedes) [inline web/multinode/src/app/store/nodes.ts:110]: It seems that you want to ensure that nodes are an Array to go into this conditional because you're calling an Array method then `if (Array.isArray(nodes))` is enough.

Comment (ptdatta) [reply web/multinode/src/app/store/nodes.ts]: Actually one test was failing for if the nodes array become `undefined` So I wanted to make sure. Removed the nodes check.

**Review by profclems — CHANGES_REQUESTED** [review]
This is the report from the CI regarding the lint failure: https://build.dev.storj.tools/blue/rest/organizations/jenkins/pipelines/storj/branches/PR-7146/runs/8/nodes/112/log/?start=0

**Comment by ptdatta** [comment]
> This is the report from the CI regarding the lint failure: https://build.dev.storj.tools/blue/rest/organizations/jenkins/pipelines/storj/branches/PR-7146/runs/8/nodes/112/log/?start=0

Fixed.✅

**Review by ifraixedes — APPROVED** [review]
